pycbc/events/single.py

```diff
@@ -31,31 +65,69 @@ def from_cli(cls, args, ifo):
         return cls(
            ifo, newsnr_threshold=args.single_newsnr_threshold[ifo],
            reduced_chisq_threshold=args.single_reduced_chisq_threshold[ifo],
-           fixed_ifar=args.single_fixed_ifar[ifo],
            duration_threshold=args.single_duration_threshold[ifo],
+           fixed_ifar=args.single_fixed_ifar,
+           fit_file=args.single_fit_file,
+           sngl_ifar_est_dist=args.sngl_ifar_est_dist[ifo]
            )
 
-    def check(self, triggers, data_reader):
+    def check(self, trigs, data_reader):
         """ Look for a single detector trigger that passes the thresholds in
         the current data.
         """
-        if len(triggers['snr']) == 0:
-            return None
-
-        i = triggers['snr'].argmax()
-        # This uses the pycbc live convention of chisq always meaning the
-        # reduced chisq.
-        rchisq = triggers['chisq'][i]
-        nsnr = ranking.newsnr(triggers['snr'][i], rchisq)
-        dur = triggers['template_duration'][i]
-
-        if nsnr > self.newsnr_threshold and \
-                rchisq < self.reduced_chisq_threshold and \
-                dur > self.duration_threshold:
-            fake_coinc = {'foreground/%s/%s' % (self.ifo, k): triggers[k][i]
-                          for k in triggers}
+        trigsremain = True
+        if len(trigs['snr']) == 0:
+            trigsremain = False
+
+        if trigsremain:
+            # Apply cuts to trigs before clustering
+            valid_idx = (trigs['template_duration'] >
+                         self.thresholds['duration']) & \
+                        (trigs['chisq'] < self.thresholds['reduced_chisq'])
+            if not np.any(valid_idx):
+                trigsremain = False
+
+        if trigsremain:
+            cutdurchi_trigs = {k: trigs[k][valid_idx] for k in trigs}
+            nsnr_all = ranking.newsnr(cutdurchi_trigs['snr'],
+                                      cutdurchi_trigs['chisq'])
+            nsnr_idx = nsnr_all > self.thresholds['newsnr']
+            if not np.any(nsnr_idx):
+                trigsremain = False
+        if trigsremain:
+            cutall_trigs = {k: cutdurchi_trigs[k][nsnr_idx]
+                            for k in trigs}
+
+            # 'cluster' by taking the maximal newsnr value over the trigger set
+            i = nsnr_all[nsnr_idx].argmax()
+
+            # This uses the pycbc live convention of chisq always meaning the
+            # reduced chisq.
+            rchisq = cutall_trigs['chisq'][i]
+            nsnr = ranking.newsnr(cutall_trigs['snr'][i], rchisq)
+            dur = cutall_trigs['template_duration'][i]
+
+        if trigsremain and nsnr > self.thresholds['newsnr'] and \
+                dur > self.thresholds['duration'] and \
+                rchisq < self.thresholds['reduced_chisq']:
```

Comment:
At this point the trigger already passed the newsnr, duration and rchisq thresholds right? Can you avoid checking the thresholds again?

Reply:
Thats right - I will remove the check here, but I have just realised that I should be doing an SNR cut in the first round of cuts so that newsnr is only calculated if SNR is above threshold